src/aleph/http/multipart.clj

```diff
@@ -162,42 +170,87 @@
     (let [req' (.finalizeRequest encoder)]
       [req' (when (.isChunked encoder) encoder)])))
 
+(defrecord MultipartChunk [part-name
+                           content
+                           name
+                           charset
+                           mime-type
+                           transfer-encoding
+                           memory?
+                           file?
+                           file
+                           size
+                           ^ReferenceCounted raw-http-data]
+  ReferenceCounted
+  (refCnt [_]
+    (.refCnt raw-http-data))
+  (retain [_]
+    (.retain raw-http-data))
+  (retain [_ increment]
+    (.retain raw-http-data increment))
+  (^ReferenceCounted touch [_]
+    (.touch raw-http-data))
+  (^ReferenceCounted touch [_ ^Object hint]
+    (.touch raw-http-data hint))
+  (release [_]
+    (.release raw-http-data))
+  (release [_ decrement]
+    (.release raw-http-data decrement)))
+
 (defmulti http-data->map
   (fn [^InterfaceHttpData data]
     (.getHttpDataType data)))
 
 (defmethod http-data->map InterfaceHttpData$HttpDataType/Attribute
   [^Attribute attr]
   (let [content (.getValue attr)]
-    {:part-name (.getName attr)
-     :content content
-     :name nil
-     :charset (-> attr .getCharset .toString)
-     :mime-type nil
-     :transfer-encoding nil
-     :memory? (.isInMemory attr)
-     :file? false
-     :file nil
-     :size (count content)}))
+    (map->MultipartChunk
+     {:part-name (.getName attr)
+      :content content
+      :name nil
+      :charset (-> attr .getCharset .toString)
+      :mime-type nil
+      :transfer-encoding nil
+      :memory? (.isInMemory attr)
+      :file? false
+      :file nil
+      :size (count content)
+      :raw-http-data attr})))
 
 (defmethod http-data->map InterfaceHttpData$HttpDataType/FileUpload
   [^FileUpload data]
   (let [memory? (.isInMemory data)]
-    {:part-name (.getName data)
-     :content (when memory?
-                (bs/to-input-stream (netty/acquire (.content data))))
-     :name (.getFilename data)
-     :charset (-> data .getCharset .toString)
-     :mime-type (.getContentType data)
-     :transfer-encoding (.getContentTransferEncoding data)
-     :memory? memory?
-     :file? true
-     :file (when-not memory? (.getFile data))
-     :size (.length data)}))
+    (map->MultipartChunk
+     {:part-name (.getName data)
+      :content (when memory?
+                 (bs/to-input-stream (netty/acquire (.content data))))
+      :name (.getFilename data)
+      :charset (-> data .getCharset .toString)
+      :mime-type (.getContentType data)
+      :transfer-encoding (.getContentTransferEncoding data)
+      :memory? memory?
+      :file? true
+      :file (when-not memory? (.getFile data))
+      :size (.length data)
+      :raw-http-data data})))
 
 (defn- read-attributes [^HttpPostRequestDecoder decoder parts]
```

Comment:
Can we rename this private fn? Nothing about it suggests it "reads attributes" to me.

Reply:
It's named that way because it reads `io.netty.handler.codec.http.multipart.MixedAttribute` [1] or `io.netty.handler.codec.http.multipart.MixedFileUpload` [2].

I will look for a better name.


[1] : https://github.com/netty/netty/blob/4.1/codec-http/src/main/java/io/netty/handler/codec/http/multipart/MixedAttribute.java#L27
[2] : https://github.com/netty/netty/blob/4.1/codec-http/src/main/java/io/netty/handler/codec/http/multipart/MixedFileUpload.java